### PR TITLE
Add an example of column generation

### PR DESCRIPTION
**Pull request by dourouc05**
Follows #2004

### CODE REVIEW

**Comment by dourouc05** [comment]
Here is an update example, with a lot more description (the problem to solve, a general description of the technique), but I did not write a column-generation tutorial. However, I did not change the dividing in functions, as there is just a main function and the pricing problem: merging the two would make things less clear, I fear.

**Comment by matbesancon** [comment]
@dourouc05 cool that you're using it. Instead of a repo I will archive both the post and code in a reproducible state (with Manifest and all) to cite something stable

File: examples/cutting_stock_column_generation.jl [hunk 1]
```diff
@@ -0,0 +1,203 @@
+# Based on https://matbesancon.github.io/post/2018-05-25-colgen2/
+# and https://github.com/matbesancon/column_generation_jump
+
+using JuMP, GLPK
+using SparseArrays
+const MOI = JuMP.MathOptInterface
```
Comment (blegat) [inline examples/cutting_stock_column_generation.jl]: This is already exported by JuMP

**Comment by matbesancon** [comment]
@dourouc05 you can use `10.5281/zenodo.3329388` as DOI or:

```bibtex
@misc{mathieu_besancon_2019_3329389,
  author       = {Mathieu Besançon},
  title        = {{A column generation example in Julia and JuMP}},
  month        = jul,
  year         = 2019,
  doi          = {10.5281/zenodo.3329388},
  url          = {https://doi.org/10.5281/zenodo.3329388}
}
```

**Comment by dourouc05** [comment]
I did not add the full BibTeX entry, as it is easy enough to find from Zenodo. Is that good enough for you?

**Comment by matbesancon** [comment]
Yup all good thanks

File: examples/cutting_stock_column_generation.jl [hunk 2]
```diff
@@ -0,0 +1,194 @@
+# Based on http://doi.org/10.5281/zenodo.3329388
```
Comment (mlubin) [inline examples/cutting_stock_column_generation.jl]: Please add the standard license header before the zenodo link.

File: examples/cutting_stock_column_generation.jl [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        # and go on waltzing one more time with the pricing problem.
+        ncols += 1
+        patterns = hcat(patterns, new_pattern)
+
+        # One new variable.
+        new_var = @variable(m, [ncols], base_name="θ", lower_bound=0)
+        push!(θ, new_var[ncols])
+
+        # Update the objective function.
+        set_objective_function(m, objective_function(m)
```
Comment (mlubin) [inline examples/cutting_stock_column_generation.jl]: This is ugly. Let's consider implementing `set_objective_coefficient` first.

File: examples/cutting_stock_column_generation.jl [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    # Then, generate new patterns, based on the dual information.
+    while ncols - n <= max_gen_cols # Generate at most max_gen_cols columns.
+        if ! has_duals(m)
+            break
+        end
+
+        new_pattern = try
+            solve_pricing(dual.(demand_satisfaction), maxwidth, widths, rollcost, demand, prices)
+        catch
+            # At the final iteration, GLPK has dual values, but at least one of them is 0.0, and thus GLPK crashes.
```
Comment (mlubin) [inline examples/cutting_stock_column_generation.jl]: Why does a dual value of 0.0 cause GLPK to crash? That shouldn't happen.

Comment (dourouc05) [reply examples/cutting_stock_column_generation.jl]: Actually, it's more GPLK.jl that gives an error (Cbc.jl too): 

```
ERROR: LoadError: AssertionError: dual >= 0.0
Stacktrace:
 [1] __assert_dual_sense__ at C:\Users\…\.julia\packages\LinQuadOptInterface\ZMx9f\src\solve.jl:210 [inlined]
 [2] get at C:\Users\…\.julia\packages\LinQuadOptInterface\ZMx9f\src\solve.jl:217 [inlined]
 [3] get(::MathOptInterface.Bridges.LazyBridgeOptimizer{GLPK.Optimizer,MathOptInterface.Utilities.UniversalFallback{MathOptInterface.Bridges.AllBridgedConstraints{Float64}}}, ::MathOptInterface.ConstraintDual, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}}) at C:\Users\…\.julia\packages\MathOptInterface\C3lip\src\Bridges\bridgeoptimizer.jl:236
 [4] get(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}, ::MathOptInterface.ConstraintDual, ::MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}}) at C:\Users\…\.julia\packages\MathOptInterface\C3lip\src\Utilities\cachingoptimizer.jl:453
 [5] _moi_get_result(::MathOptInterface.Utilities.CachingOptimizer{MathOptInterface.AbstractOptimizer,MathOptInterface.Utilities.UniversalFallback{JuMP._MOIModel{Float64}}}, ::MathOptInterface.ConstraintDual, ::Vararg{Any,N} where N) at C:\Users\…\.julia\dev\JuMP\src\JuMP.jl:618
 [6] get(::Model, ::MathOptInterface.ConstraintDual, ::ConstraintRef{Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}},ScalarShape}) at C:\Users\…\.julia\dev\JuMP\src\JuMP.jl:649
 [7] _constraint_dual(::ConstraintRef{Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}},ScalarShape}) at C:\Users\…\.julia\dev\JuMP\src\constraints.jl:545
 [8] dual(::ConstraintRef{Model,MathOptInterface.ConstraintIndex{MathOptInterface.ScalarAffineFunction{Float64},MathOptInterface.GreaterThan{Float64}},ScalarShape}) at C:\Users\…\.julia\dev\JuMP\src\constraints.jl:538
 [9] _broadcast_getindex_evalf at .\broadcast.jl:625 [inlined]
 [10] _broadcast_getindex at .\broadcast.jl:598 [inlined]
 [11] getindex at .\broadcast.jl:558 [inlined]
 [12] copyto_nonleaf!(::Array{Float64,1}, ::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1},Tuple{Base.OneTo{Int64}},typeof(dual),Tuple{Base.Broadcast.Extruded{Array{ConstraintRef{Model,C,Shape} where Shape<:AbstractShape where C,1},Tuple{Bool},Tuple{Int64}}}}, ::Base.OneTo{Int64}, ::Int64, ::Int64) at .\broadcast.jl:982
 [13] copy at .\broadcast.jl:836 [inlined]
 [14] materialize at .\broadcast.jl:798 [inlined]
 [15] #example_cutting_stock#3(::Int64, ::typeof(example_cutting_stock)) at C:\Users\…\Desktop\a.jl:126
 [16] example_cutting_stock() at C:\Users\…\Desktop\a.jl:88
 [17] top-level scope at C:\Users\…\Desktop\a.jl:175
```

**Comment by dourouc05** [comment]
I'm messing up with the rebase, I'll open a new PR.

**Pull request closed without merging**